Add documentation on how to use Tracy and Perfetto (#11484)
godotengine/godot-docs: engine_details/development/debugging/profiling/tracing_profilers.rst

Configuration
~~~~~~~~~~~~~

Perfetto requires a configuration file to tell it which events to track.

Create a file called ``godot.config`` inside of the ``perfetto`` directory
with this content:

.. code-block:: text

    # Trace for 10 seconds.
    duration_ms: 10000

    buffers {
        size_kb: 32768
        fill_policy: RING_BUFFER
    }

    # Write to file every 1 second to prevent overflowing the buffer.
    write_into_file: true
    file_write_period_ms: 1000

    # Track events in the "godot" category.
    data_sources {
        config {
            name: "track_event"
            track_event_config {
            enabled_categories: "godot"
            }
        }
    }

Record a trace
~~~~~~~~~~~~~~

Finally, launch your game on an Android device using the export templates you
built earlier.

When you're ready to record a trace (for example, when you've hit the part of
your game that is exhibiting performance issues), you can use this script that
comes with the Perfetto source code:

.. code-block:: shell

    cd perfetto
    ./tools/record_android_trace -c godot.config

This will record for 10 seconds (per the configuration), or until you press

> CTRL-C.

Author: Thanks! I've applied all your suggestions :-)

Examining the trace
~~~~~~~~~~~~~~~~~~~

As soon as that script exits, it will launch the Perfetto UI in a web browser.

To see the Godot events, expand the row for your application by clicking on its
Android "Unique Name" (Perfetto will also include some events from system
services in the trace).

.. image:: img/cpp_profiler_perfetto.webp

Then you can use the ``WASD`` keys to navigate the graph:

- Press ``A`` or ``D`` to navigate forward or backward along the timeline

> - Press ``W`` or ``S`` to zoom in or out

Reviewer: ```suggestion
- Press :kbd:`A` or :kbd:`D` to navigate forward or backward along the timeline
- Press :kbd:`W` or :kbd:`S` to zoom in or out
```

.. _doc_tracing_profilers:

Tracing Profilers
=================

> Godot currently supports two tracing profilers: Tracy and Perfetto.

Reviewer 2: ```suggestion
Godot currently supports two tracing profilers: `Tracy <https://github.com/wolfpld/tracy>`__ and `Perfetto <https://perfetto.dev>`__.
```

In order to use either of them, you'll need to build the engine from source.
If you've never done this before, please read
:ref:`these docs <doc_compiling_index>` for the platform you want to profile on.
You'll need to perform the same steps here, but with the addition of a single
extra argument for ``scons``.

.. _doc_tracy_profiler:

> Tracy for Windows, Linux and MacOS

Reviewer: ```suggestion
Tracy for Windows, Linux, and MacOS
```

----------------------------------

Tracy is an Open Source profiler that runs on a wide variety of platforms,

> including Windows, Linux and MacOS. While it is primarily a tracing profiler,

Reviewer: ```suggestion
including Windows, Linux, and MacOS. While it is primarily a tracing profiler,
```

it can also periodically sample data like a
:ref:`sampling profiler <doc_sampling_profilers>`, giving some of the benefits
of both approaches.

Build Godot with Tracy support
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

First, clone the latest version of the Tracy source code ("0.13.0" at the

> time of this writing) using Git:

Author: "at the time of this writing" is a set phrase, although, I've also seen "at the time of writing" (with "this" removed) which I guess could be a good compromise :-)

This will create a ``tracy`` directory - you can place this anywhere.

Next, build the release templates for your platform using ``scons``, but adding
the ``profiler_path=path/to/tracy`` argument with the real path to the ``tracy``
directory, as well as ``debug_symbols=yes`` to allow Tracy's sampling features
to work.

.. note::

    You don't have to build release templates, you could also build debug
    templates, or even the editor. However, it's generally recommended to
    profile release templates, because that is the version your players will
    use, and it will perform differently than other types of builds.

For example, to build release templates for Windows:

.. code-block:: shell

    scons platform=windows target=template_release debug_symbols=yes profiler_path=path/to/tracy

Get the Tracy "server"
~~~~~~~~~~~~~~~~~~~~~~

In Tracy terminology, the application you are profiling is the "client", and
the one receiving the data is the "server".

If you are on Windows, you can download a pre-built ``tracy-profiler.exe``
from the Tracy `releases page <https://github.com/wolfpld/tracy/releases>`_.

> However, if you're on Linux or MacOS, you'll need to build it yourself.

Reviewer 2: I actually had trouble building tracy at first, but later found out both `brew` and `nix` supply pre-built tracy binaries. I would not recommend building it yourself by default.

```suggestion
If you're on Linux or macOS, you may be able to get a Tracy binary from a package manager.
For example, on macOS, you can install Tracy using homebrew:

.. code-block:: shell

    brew install tracy

Alternatively, you can build the Tracy server yourself, as explained below.
```

Reviewer 3: It won't work if installed `tracy` version is not matching the client (or at least protocol version). Current `brew` version (0.11.1) doesn't work with current GitHub release client (0.13).

Author: Thanks!

In my latest update, I've mentioned that you might be able to find a pre-built binary from a package manager, but that if you do, you should be careful to match the version

Build the Tracy server from source
++++++++++++++++++++++++++++++++++

The full details can be found in the
`Tracy manual <https://github.com/wolfpld/tracy/releases/latest/download/tracy.pdf>`_,
but here is the TL;DR:

.. code-block:: shell

    # On Linux, Tracy uses Wayland by default, so if you use X11 add -DLEGACY=1
    cmake -B profiler/build -S profiler -DCMAKE_BUILD_TYPE=Release
    cmake --build profiler/build --config Release --parallel

If it builds successfully, the ``tracy-profiler`` executable (or ``tracy-profiler.exe``
on Windows) will be in ``profiler/build``.

Record a trace
~~~~~~~~~~~~~~

First, export your game using the release templates you built above, and run it.

Then, launch the Tracy server - you'll see something like this:

.. image:: img/cpp_profiler_tracy_start.webp

When you're ready to record a trace (for example, when you've hit the part of
your game that is exhibiting performance issues), press the "Connect" button.

> You'll immediately see tracing data start coming in:

Reviewer 2: ```suggestion
Then, launch the Tracy server - you'll see something like this:

.. image:: img/cpp_profiler_tracy_start.webp

If you press "Connect" now, you can profile the launch of Godot. However,
if you want to profile something specific (for example, a part of your game
that is exhibiting performance issues), wait before pressing the button.

Now, export your game using the release templates you built above, and run it.
As soon as the both the game are running, and you have pressed the "Connect"
button in Tracy, you'll see data coming in:
```

.. image:: img/cpp_profiler_tracy_recording.webp

When you think you've gathered enough data, press the "Stop" button. If you
clicked somewhere and the box with the "Stop" button disappeared, you can
click the top-left most icon to bring it back.

Examining the trace
~~~~~~~~~~~~~~~~~~~

Here are some of the basic controls:

- Zoom in/out with the mouse wheel
- Right click and drag to move forward/backward on the timeline
- In the top bar, click the left and right arrow buttons by "Frames" to move a single frame on the timeline

To learn more, see the
`Tracy manual <https://github.com/wolfpld/tracy/releases/latest/download/tracy.pdf>`_.

Perfetto for Android
--------------------

Perfetto is the default tracing system for Android. In fact, its system tracing
service has been built into the platform since Android 9.

Build Godot with Perfetto support
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

First, clone the latest version of the Perfetto source code ("53.0" at the

> time of this writing) using Git:

Reviewer: ```suggestion
time of writing this) using Git:
```

        config {
            name: "track_event"
            track_event_config {
                enabled_categories: "godot"
            }
        }
    }

Record a trace
~~~~~~~~~~~~~~

Finally, launch your game on an Android device using the export templates you
built earlier.

When you're ready to record a trace (for example, when you've hit the part of
your game that is exhibiting performance issues), you can use this script that
comes with the Perfetto source code:

.. code-block:: shell

    cd perfetto
    ./tools/record_android_trace -c godot.config

This will record for 10 seconds (per the configuration), or until you press
:kbd:`Ctrl + C`.

Examining the trace
~~~~~~~~~~~~~~~~~~~

As soon as that script exits, it will launch the Perfetto UI in a web browser.

To see the Godot events, expand the row for your application by clicking on its
Android "Unique Name" (Perfetto will also include some events from system
services in the trace).

.. image:: img/cpp_profiler_perfetto.webp

Then you can use the ``WASD`` keys to navigate the graph:

- Press :kbd:`A` or :kbd:`D` to navigate forward or backward along the timeline
- Press :kbd:`W` or :kbd:`S` to zoom in or out

> You'll probably need to zoom a bit before your able to see the individual

Reviewer: ```suggestion
You'll probably need to zoom a bit before you're able to see the individual
```

:ref:`these docs <doc_compiling_index>` for the platform you want to profile on.
You'll need to perform the same steps here, but with the addition of a single
extra argument for ``scons``.

.. _doc_tracy_profiler:

Tracy for Windows, Linux, and macOS
-----------------------------------

Tracy is an Open Source profiler that runs on a wide variety of platforms,
including Windows, Linux, and macOS. While it is primarily a tracing profiler,
it can also periodically sample data like a
:ref:`sampling profiler <doc_sampling_profilers>`, giving some of the benefits
of both approaches.

Build Godot with Tracy support
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

First, clone the latest version of the Tracy source code ("0.13.0" at the
time of writing) using Git:

.. code-block:: shell

    git clone -b v0.13.0 --single-branch https://github.com/wolfpld/tracy.git

This will create a ``tracy`` directory - you can place this anywhere.

Next, build the release templates for your platform using ``scons``, but adding
the ``profiler_path=path/to/tracy`` argument with the real path to the ``tracy``
directory, as well as ``debug_symbols=yes`` to allow Tracy's sampling features

> to work.

Reviewer 2: > Hrm, I wasn't aware of this change. I'm not sure I like it - having to specify two arguments instead of one seems less friendly :-/

I voiced the same reservation, but was outvoted :D

However, if you're on Linux or macOS, you'll either need to find a pre-built
binary from a package manager (like ``brew`` or ``nix``), or build it from
source yourself.

.. note::

    If you do use a pre-built binary, be sure to use the same version that
    you used when building Godot.

Build the Tracy server from source
++++++++++++++++++++++++++++++++++

The full details can be found in the
`Tracy manual <https://github.com/wolfpld/tracy/releases/latest/download/tracy.pdf>`_,
but here is the TL;DR:

.. code-block:: shell

    # On Linux, Tracy uses Wayland by default, so if you use X11 add -DLEGACY=1
    cmake -B profiler/build -S profiler -DCMAKE_BUILD_TYPE=Release
    cmake --build profiler/build --config Release --parallel

Record a trace
~~~~~~~~~~~~~~

Launch the Tracy server - you'll see something like this:

.. image:: img/cpp_profiler_tracy_start.webp

If you press "Connect" now, you can profile the launch of Godot. However,
if you want to profile something specific (for example, a part of your game
that is exhibiting performance issues), wait before pressing the button.

Now, export your game using the release templates you built above, and run it.

> As soon as both are running, and you have pressed the "Connect" button in

Author: Perhaps we could add a "Troubleshooting" section later? I don't think we should get into too much detail in the middle of the steps

button in Tracy, you'll see data coming in:

.. image:: img/cpp_profiler_tracy_recording.webp

When you think you've gathered enough data, press the "Stop" button. If you
clicked somewhere and the box with the "Stop" button disappeared, you can
click the top-left most icon to bring it back.

Examining the trace
~~~~~~~~~~~~~~~~~~~

Here are some of the basic controls:

- Zoom in/out with the mouse wheel
- Right click and drag to move forward/backward on the timeline
- In the top bar, click the left and right arrow buttons by "Frames" to move a single frame on the timeline

To learn more, see the
`Tracy manual <https://github.com/wolfpld/tracy/releases/latest/download/tracy.pdf>`_.

> Perfetto for Android

Author: I don't think the way Godot has integrated it (at least so far) really supports Windows and Mac. It's setup to use the "system" backend, which requires connecting with Perfetto's system daemon which I think is only Linux and Android. If we added support for the "in process" backend we could add docs for how to use that later? The steps for using it would be somewhat different too

.. _doc_tracing_profilers:

Tracing Profilers
=================

Godot currently supports two tracing profilers:
`Tracy <https://github.com/wolfpld/tracy>`__ and `Perfetto <https://perfetto.dev>`__.

In order to use either of them, you'll need to build the engine from source.
If you've never done this before, please read
:ref:`these docs <doc_compiling_index>` for the platform you want to profile on.
You'll need to perform the same steps here, but with some additional arguments
for ``scons``.

.. _doc_tracy_profiler:

Tracy for Windows, Linux, and macOS
-----------------------------------

Tracy is an Open Source profiler that runs on a wide variety of platforms,
including Windows, Linux, and macOS. While it is primarily a tracing profiler,
it can also periodically sample data like a
:ref:`sampling profiler <doc_sampling_profilers>`, giving some of the benefits
of both approaches.

Build Godot with Tracy support
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

First, clone the latest version of the Tracy source code ("0.13.0" at the
time of writing) using Git:

.. code-block:: shell

>     git clone -b v0.13.0 --single-branch https://github.com/wolfpld/tracy.git

Reviewer 4: Tracy has relatively frequent releases so this will get out of date quickly, and we don't want to have to update it regularly (which would add churn and also invalidate translations), so I'd suggest looking for an alternative that doesn't hardcode a version number.

You can use https://github.com/wolfpld/tracy/releases/latest to link to the latest tag, but that's not usable for the git clone command.
